### PR TITLE
Move to new Dynamic Parameters

### LINKED ISSUE #1581: Set range of parameters in GUI
### Description of the task

Currently when we create widgest in the GUI for DynamicParams, we just allow the widget to use the full range of whatever type the parameters is (eg. min_int to max_int). Once the new parameter system is up and running with `NumericParameter` and `EnumeratedParameter`, we should be able to automatically get the range of the parameters and set this in the GUI.

### Acceptance criteria

- [ ] Set the range of each DynamicParameter in the GUI widgets

### Blocked By

#1298

### LINKED ISSUE #1892: Replace existing dynamic parameter system with new system
### Description of the task

Replace existing dynamic parameters system with the new one.

Note: The names of getters in the new system have been slightly modified to be consistent in convention; ie. old system const accessor to a parameter named `EnemyProximityImportance` would be changed from `EnemyProximityImportance()` -> `enemyProximityImportance()`

### Acceptance criteria

- [ ] Use the new system in the code base
- [ ] Delete old system
- [ ] Rename new system top level config class from `ThunderbotsConfigNew` to `ThunderbotsConfig`

### Blocked By

- #1891

### PR DESCRIPTION
**Pull request by akhilveeraghanta**
### Description
Moving to new system! Just a drop in replacement, didn't add or remove any params.

### Testing Done
All unit tests pass, simulator and visualizer run correctly

### Resolved Issues
resolves #1581
resolves #1892

### Length Justification
Not much review is really required here, and its hard to split this pr. 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/shared/parameter_v2/config_definitions/passing_config.yaml
Comment (tarakong01) [inline src/shared/parameter_v2/config_definitions/passing_config.yaml]: Just to clarify, is the "robot near the pass receiver" referring to an enemy robot?

File: src/shared/parameter_v2/config_definitions/shoot_goal_tactic_config.yaml
Comment (tarakong01) [inline src/shared/parameter_v2/config_definitions/shoot_goal_tactic_config.yaml:8]: I think you forgot to add "steal the ball" to the end of the line!

**Review by tarakong01 — COMMENTED** [review]
Added a couple comments where I needed a bit of clarification and found a typo! :)

**Review by jsn0716 — CHANGES_REQUESTED** [review]
Looking good, thanks for doing the replacement. Looks like some tests are failing CI, including `cpp_config_test.cpp`. This test seems to be failing from still using `ThunderbotsConfigNew`, and the allowed acronyms change in `case_conversion.py` as the yaml generated from the test doens't follow this rule atm

File: src/shared/parameter_v2/generation_scripts/cpp_parameter.py
Comment (jonathanlew) [reply src/shared/parameter_v2/generation_scripts/cpp_parameter.py:136]: what if we made it `getParameterName`?

File: src/shared/parameter_v2/config_definitions/corner_kick_play_config.yaml
Comment (jonathanlew) [inline src/shared/parameter_v2/config_definitions/corner_kick_play_config.yaml:1]: What if we put all of these into a playconfig?

**Review by jonathanlew — CHANGES_REQUESTED** [review]
good to see this coming together

File: src/software/simulated_tests/simulated_test_fixture.cpp
Comment (itsarune) [inline src/software/simulated_tests/simulated_test_fixture.cpp]: why is this just commented out and not deleted altogether?

**Review by itsarune — COMMENTED** [review]
Nothing much more to add for me, just something I wanted some clarification on

File: src/shared/parameter_v2/config_definitions/passing_config.yaml
Comment (akhilveeraghanta) [reply src/shared/parameter_v2/config_definitions/passing_config.yaml]: yeah it is, i've updated the description to be more clear

File: src/shared/parameter_v2/config_definitions/shoot_goal_tactic_config.yaml
Comment (akhilveeraghanta) [reply src/shared/parameter_v2/config_definitions/shoot_goal_tactic_config.yaml:8]: good catch! added

File: src/shared/parameter_v2/config_definitions/test/example.yaml
Comment (akhilveeraghanta) [reply src/shared/parameter_v2/config_definitions/test/example.yaml]: This will be handled by https://github.com/UBC-Thunderbots/Software/issues/1738 when we actually document dynamic parameters

File: src/shared/parameter_v2/generation_scripts/cpp_parameter.py
Comment (akhilveeraghanta) [reply src/shared/parameter_v2/generation_scripts/cpp_parameter.py:46]: the min max are required for every numeric value, the schema will complain if someone doesn't provide it

So it shouldn't be needed/used anymore

File: src/shared/parameter_v2/generation_scripts/cpp_parameter.py
Comment (akhilveeraghanta) [reply src/shared/parameter_v2/generation_scripts/cpp_parameter.py:136]: i've gone ahead and fixed everything to be camel_case w/ get infront

File: src/shared/parameter_v2/generation_scripts/dynamic_parameter_schema.py
Comment (akhilveeraghanta) [reply src/shared/parameter_v2/generation_scripts/dynamic_parameter_schema.py:10]: float is relevant for c params so its still technically a supported type

File: src/software/parameter/config.h
Comment (akhilveeraghanta) [reply src/software/parameter/config.h:31]: its still technically allowed for c parameters, and we _could_ use it for cpp in which case we should add it back to the variant

currently no param uses it so i've removed it from the `config.h`

File: src/shared/parameter_v2/BUILD
Comment (akhilveeraghanta) [reply src/shared/parameter_v2/BUILD:3]: it wasn't being used in software, only for test in the `parameter_v2` folder

File: src/shared/parameter_v2/config_definitions/corner_kick_play_config.yaml
Comment (akhilveeraghanta) [reply src/shared/parameter_v2/config_definitions/corner_kick_play_config.yaml:1]: yeah that would be a good idea, in this PR im not really reorganizing the params

will do this in the next pass

File: src/software/simulated_tests/simulated_test_fixture.cpp
Comment (akhilveeraghanta) [reply src/software/simulated_tests/simulated_test_fixture.cpp]: ah really good catch! uncommented

**Review by jsn0716 — DISMISSED** [review]
LGTM 💯

File: src/shared/parameter_v2/config_definitions/passing_config.yaml
Comment (kelvinkoon) [inline src/shared/parameter_v2/config_definitions/passing_config.yaml]: Nit: I think weight is a noun and weigh is a verb.

File: src/shared/parameter_v2/config_definitions/passing_config.yaml
Comment (kelvinkoon) [inline src/shared/parameter_v2/config_definitions/passing_config.yaml:45]: Some values in the `handheld_controller_config` such as `max_linear_speed` don't include units while others do. Is there a standard notation for if we want to include units in the names?

File: src/shared/parameter_v2/config_definitions/passing_config.yaml
Comment (kelvinkoon) [inline src/shared/parameter_v2/config_definitions/passing_config.yaml:80]: What is the value this takes if the enemy does not recognize we're passing / does not try to intercept?

File: src/shared/parameter_v2/config_definitions/shoot_or_pass_play_config.yaml
Comment (kelvinkoon) [inline src/shared/parameter_v2/config_definitions/shoot_or_pass_play_config.yaml]: Is this referring to the parameter above? If so, update `absMinPassScore` to `abs_min_pass_score`.

**Review by kelvinkoon — CHANGES_REQUESTED** [review]
💯

File: src/software/backend/radio_backend.cpp
Comment (itsarune) [inline src/software/backend/radio_backend.cpp:9]: The name "SslCommunicationConfig" is a bit awkward. is there a way to keep it the same as before?

File: src/shared/parameter_v2/config_definitions/full_system_main_command_line_args.yaml
Comment (akhilveeraghanta) [reply src/shared/parameter_v2/config_definitions/full_system_main_command_line_args.yaml:32]: well it fails because there are no protologs to replay, should be a runtime error

File: src/shared/parameter_v2/config_definitions/network_config.yaml
Comment (akhilveeraghanta) [reply src/shared/parameter_v2/config_definitions/network_config.yaml:16]: This should still be `ifconfig` because you are running in ubuntu.

File: src/shared/parameter_v2/config_definitions/passing_config.yaml
Comment (akhilveeraghanta) [reply src/shared/parameter_v2/config_definitions/passing_config.yaml:80]: It would be 3 seconds in this case

File: src/shared/parameter_v2/config_definitions/goalie_tactic_config.yaml
Comment (akhilveeraghanta) [reply src/shared/parameter_v2/config_definitions/goalie_tactic_config.yaml]: removed, its not really relevant, should be obvious from the min/max

File: src/shared/parameter_v2/config_definitions/passing_config.yaml
Comment (akhilveeraghanta) [reply src/shared/parameter_v2/config_definitions/passing_config.yaml]: fixed

File: src/shared/parameter_v2/config_definitions/passing_config.yaml
Comment (akhilveeraghanta) [reply src/shared/parameter_v2/config_definitions/passing_config.yaml:45]: This will be handled by this #1738 , we should have a standard to inlude units in the name or atleast in the description

File: src/shared/parameter_v2/config_definitions/shoot_or_pass_play_config.yaml
Comment (akhilveeraghanta) [reply src/shared/parameter_v2/config_definitions/shoot_or_pass_play_config.yaml]: fixed

File: src/software/backend/radio_backend.cpp
Comment (akhilveeraghanta) [reply src/software/backend/radio_backend.cpp:9]: Its actually a convention to treat acronyms as words and use camelCase which makes things much easier to work with. Jason mentioned hes seen this convention in use before

We would need to create exceptions for `Ai, Ssl, Id`, its _slightly_ annoying to create these exceptions but we can do it.

https://stackoverflow.com/questions/15526107/acronyms-in-camelcase

Thoughts @itsarune  @jonathanlew  @tarakong01 @jsn0716 @kelvinkoon ?

File: src/shared/parameter_v2/config_definitions/passing_config.yaml
Comment (kelvinkoon) [reply src/shared/parameter_v2/config_definitions/passing_config.yaml:80]: Is there any significance to setting the value to 0.4 instead of 3.0 in that case?

Comment (akhilveeraghanta) [reply src/shared/parameter_v2/config_definitions/passing_config.yaml:80]: not sure, we haven't really done a sensitivity study on these values

Comment (kelvinkoon) [reply src/shared/parameter_v2/config_definitions/passing_config.yaml:80]: I was under the impression the `value` field was a default value which is why I asked. If we're fine with it, then I don't have issues either but it might be worth noting down.

**Review by kelvinkoon — DISMISSED** [review]
🆕

File: src/shared/parameter_v2/config_definitions/passing_config.yaml
Comment (akhilveeraghanta) [reply src/shared/parameter_v2/config_definitions/passing_config.yaml:80]: The value field is the default, we have a _ton_ of dynamic params rn, some of which we probably dont even need. 
So we can look at the effects of this param and see if it actually does anything/tweak those 3 values or remove it altogether if it doesn't

File: src/software/backend/radio_backend.cpp
Comment (kelvinkoon) [reply src/software/backend/radio_backend.cpp:9]: ATM, I prefer the camelCase b/c the other similar configurations (eg. `NetworkConfig` and `SensorFusionConfig`) are also camelCase'd.

File: src/shared/parameter_v2/config_definitions/passing_config.yaml
Comment (kelvinkoon) [reply src/shared/parameter_v2/config_definitions/passing_config.yaml:80]: I feel like it makes more sense to set the default to 3.0 b/c semantically, we begin the game not knowing if they are capable of interception or not. 
Agreed, it's not a priority right now but would be nice to go through later on.

Comment (akhilveeraghanta) [reply src/shared/parameter_v2/config_definitions/passing_config.yaml:80]: Id argue the other way, id say its safe to assume the worst and that the enemy can _reasonably_ intercept our passes. 

And then we ease that if we are noticing that they cant

Comment (kelvinkoon) [reply src/shared/parameter_v2/config_definitions/passing_config.yaml:80]: Ah, that makes more sense when thinking that way then.

File: src/software/backend/radio_backend.cpp
Comment (jsn0716) [reply src/software/backend/radio_backend.cpp:9]: I definitely prefer the convention of treating acronyms as words. Consider a case where you have to write an acronym at the beginning in camel case with first letter lower case (for our code, function names), where keeping the acronyms in upper case conflicts with the first letter lower case style. While this case doesn't specifically apply here, I would prefer the styles be kept consistent.

Comment (jonathanlew) [reply src/software/backend/radio_backend.cpp:9]: I think the two-letter convention is neat, but I do like eliminating exceptions so treating all acronyms as words would be preferable. So, Ssl/ssl, Ai/ai, Id/id (for pascal/camel). We need to document this in our coding styles

**Review by jsn0716 — COMMENTED** [review]
LGTM, merge conflict needs to be resolved

File: src/software/ai/hl/stp/play/free_kick_play.cpp
Comment (jonathanlew) [inline src/software/ai/hl/stp/play/free_kick_play.cpp:59]: are we consistent across the codebase on AI vs Ai?

**Review by jonathanlew — COMMENTED** [review]
looks good except for making sure we're consistent with the acronym naming convention across the codebase. This should be documented

File: src/software/backend/radio_backend.cpp
Comment (itsarune) [reply src/software/backend/radio_backend.cpp:9]: Consistency makes sense, so keeping it is good

**Review by itsarune — COMMENTED** [review]
looks good, will approve after fixing merge conflict

File: src/software/ai/hl/stp/play/free_kick_play.cpp
Comment (akhilveeraghanta) [reply src/software/ai/hl/stp/play/free_kick_play.cpp:59]: created a ticket here: https://github.com/UBC-Thunderbots/Software/issues/1939

**Review by jonathanlew — APPROVED** [review]
LGTM pending CI

**Review by itsarune — APPROVED** [review]

**Pull request merged**
